gson/src/test/java/com/google/gson/TypeAdapterTest.java

```diff
@@ -28,23 +28,34 @@
 public class TypeAdapterTest {
   @Test
   public void testNullSafe() throws IOException {
-    TypeAdapter<String> adapter =
-        new TypeAdapter<String>() {
-          @Override
-          public void write(JsonWriter out, String value) {
-            throw new AssertionError("unexpected call");
-          }
-
-          @Override
-          public String read(JsonReader in) {
-            throw new AssertionError("unexpected call");
-          }
-        }.nullSafe();
+    TypeAdapter<String> adapter = assertionErrorAdapter.nullSafe();
 
     assertThat(adapter.toJson(null)).isEqualTo("null");
     assertThat(adapter.fromJson("null")).isNull();
   }
 
+  @Test
+  public void testNullSafe_ReturningSameInstanceOnceNullSafe() {
+    TypeAdapter<?> nullSafeAdapter = assertionErrorAdapter.nullSafe();
+
+    assertThat(nullSafeAdapter.nullSafe()).isSameInstanceAs(nullSafeAdapter);
+    assertThat(nullSafeAdapter.nullSafe().nullSafe()).isSameInstanceAs(nullSafeAdapter);
+    assertThat(nullSafeAdapter.nullSafe().nullSafe().nullSafe()).isSameInstanceAs(nullSafeAdapter);
+  }
+
+  private static final TypeAdapter<String> assertionErrorAdapter =
+      new TypeAdapter<String>() {
```

Comment:
Minor: `String` can probably be omitted here since it can be inferred
```suggestion
      new TypeAdapter<>() {
```

Reply:
I merely moved the `TypeAdapter` instance from the method to the field leaving the type parameter as it was previously in method. My rationale is that I should follow the convention and do not touch code I don't own and that can be a subject of refactoring in the future.

If you're really fine with changing from `TypeAdapter<String>`  to `TypeAdapter<>`, I'll change `<String>` to `<>`.

Reply:
> Minor: `String` can probably be omitted here since it can be inferred

My bad: I seem to have missed the line change I made accidentally. Also I removed the `assertionErrorAdapter` `TypeAdapter` type from `<String>` to `<>` -- I saw the diamond operator has been in use at least in this test class.